sdk/communication/Azure.Communication.Calling.Server/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,323 @@
+            }
+            else
+            {
+                _initialRangeSize = Constants.ContentDownloader.Partition.DefaultInitalDownloadRangeSize;
+            }
+        }
+
+        internal async Task<Response> DownloadToAsync(
+            Stream destination,
+            Uri endpoint,
+            CancellationToken cancellationToken)
+        {
+            // Wrap the download range calls in a Download span for distributed
+            // tracing
+            DiagnosticScope scope = _client._clientDiagnostics.CreateScope($"{nameof(ConversationClient)}.{nameof(ConversationClient.DownloadTo)}");
+            try
+            {
+                scope.Start();
+
+                // Just start downloading using an initial range.  If it's a
+                // small blob, we'll get the whole thing in one shot.  If it's
+                // a large blob, we'll get its full size in Content-Range and
+                // can keep downloading it in segments.
+                var initialRange = new HttpRange(0, _initialRangeSize);
+                Task<Response<Stream>> initialResponseTask =
+                    _client.DownloadStreamingAsync(
+                        endpoint,
+                        initialRange,
+                        cancellationToken);
+
+                Response<Stream> initialResponse;
+                try
+                {
+                    initialResponse = await initialResponseTask.ConfigureAwait(false);
+                }
+                catch (RequestFailedException ex) when (ex.ErrorCode == "Invalid Range")
+                {
+                    initialResponse = await _client.DownloadStreamingAsync(
+                        endpoint,
+                        range: default,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                }
+
+                // If the first segment was the entire blob, we'll copy that to
+                // the output stream and finish now
+                long initialLength = ParseResponseContentLength(initialResponse);
+                long totalLength = ParseRangeTotalLength(initialResponse);
+                if (initialLength == totalLength)
+                {
+                    await CopyToAsync(
+                        initialResponse,
+                        destination,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                    return initialResponse.GetRawResponse();
+                }
+
+                // Create a queue of tasks that will each download one segment
+                // of the blob.  The queue maintains the order of the segments
+                // so we can keep appending to the end of the destination
+                // stream when each segment finishes.
+                var runningTasks = new Queue<Task<Response<Stream>>>();
+                runningTasks.Enqueue(initialResponseTask);
```

Comment:
`initialResponseTask` is not always set (https://github.com/Azure/azure-sdk-for-net/pull/21494/files#diff-6b967692fd882eae68f3f8ecfc73741a6d48d5cf4bb7e587e607ccbaa7ff50e9R106)

Reply:
We should probably have some tests for the corner cases.

Reply:
I reassigned the `initialResponseTask` value within the `catch`.
I'll look to add more testing for this.

Reply:
Hey @RezaJooyandeh, so, I'm not able to do some unit testing for this.
At least in the way I'm mocking the HTTP response: I can mock one of the responses (The invalid range one), but for the second call I won't be able to let the Mock to answer differently.
Do you have any advice on how to accomplish such task? I was planning to try to test this with the actual server.

Reply:
> the second call I won't be able to let the Mock to answer differently.

What stops you from having it act differently? Cannot you have a stateful mock function that acts differently?

Reply:
Yeah, you were right. I figured that I could pass a list of responses and it would take one by one in order.